### PR TITLE
Convert to go mods, Separate certmgrd internals from certmgr lib code, build a usable API.

### PR DESCRIPTION
**Pull request by ferringb**
TL;DR: this splits the codebase areas of concern into a `certmgr/*` tool pathing, and a library pathing (cert/*), and then moves everything back into the actual spot it belonged in.  In the process, we wind up with a simplified 'cert.Spec' object that isn't creatable only via parsing files from disk, and also make the individual source actually readable now.

For review, I'd suggest just pulling the files in cert/* rather than looking at diff's- it's easier to read.  Note this is a usable first 'split', but I expect there to be further iterations as the code becomes more controlled/managed.

Relevant commit message meanwhile:
```
The goal here is to break out a usable API for codebases not using
certmgr spec files, while also decoupling the internals from the
certmgrd implementation.  The code was far to intertwined and overloading
of concerns- and in turn, complicating the heck out of cert.Spec.

While I'm restructuring to support this, the parsing related code,
the CLI tool, etc, all are moved into a new certmgr subpackage and
were realigned to normal standards (cobra layout for cmd's for example).

The supportive code for cert/*- primarily file/* and util/* were moved
into the areas of the codebase that makes sense, now that things are
separated out.

Renaming and cleanup is needed within the certmgr/*, but going forward
the parsing code should never be w/in the same package space as Spec
to ensure that separtion of concerns is maintained (in the process,
ensuring that certmgr is usabled as a general lib).  Note that there
are no semver API guarantees that will be made for the certmgr utility.

Use the library code instead; semver will be maintained for that.

On disk spec parsing has no compatibility breakage, but I have
moved the CA file definition out of 'authority'.

Backwards compatibility is retained, it'll just warn if you're using
the old pathing.

Backwards compatibility wasn't retained for the following:
* CHANGE_TYPE is no longer exposed for command notifications; things shouldn't
  care if it's the CA or cert that changed, it should just reimport both.
* metrics that were action orientated were dropped; they weren't particularly
  useful and they directly conflict w/ the storage interface that has been
  introduce to allow using cert.spec as a proper API.
```

### CODE REVIEW

**Comment by ferringb** [comment]
I'll push an update in a bit addressing the test complaints (not sure why I didn't trigger that locally, but oh well, that's what travis is for- a backstop at worst).

File: cert/authority.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+				"label":         a.Label,
+				"tls-remote-ca": a.RootCACert,
+				"auth-type":     "standard",
+				"auth-key":      a.AuthKey,
+			},
+		},
+	}
+}
+
+// CreateTransport converts an Authority and CertificateRequest into
```
Comment (ferringb) [inline cert/authority.go]: incomplete, fixing.

File: cert/authority.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	ident.Request = request
+	return transport.New(before, ident)
+}
+
+func (a *Authority) getRemoteCert() (*x509.Certificate, error) {
+	var tlsConfig *tls.Config
+	if a.RootCACert != "" {
+		rootCABytes, err := ioutil.ReadFile(a.RootCACert)
+		if err != nil {
+			return nil, err
```
Comment (willbuckner) [inline cert/authority.go]: It  might be useful to wrap the `err`'s in this method to add context--the error returned from `ReadFile()` may not easily trace back to context understandable by human.

File: cert/metrics.go [hunk 3]
```diff
@@ -94,56 +72,28 @@ var (
...
-			Help:      "Number of times a spec has taken action",
-		},
-		[]string{"spec_path", "change_type"},
-	)
-
-	// ActionFailedCount counts failed actions taken by a spec
-	ActionFailedCount = prometheus.NewGaugeVec(
-		prometheus.GaugeOpts{
-			Namespace: metricsNamespace,
-			Name:      "action_failed_count",
```
Comment (willbuckner) [inline cert/metrics.go]: This metric seems especially useful. I understand that it doesn't fit the new architecture, but how will we now monitor when certmgr's actions fail?

File: cert/storage/disk.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+type FileBackend struct {
+	ca   *util.CertificateFile
+	cert *util.CertificateFile
+	key  *util.File
+}
+
+// NewFileBackend creates a storage backend used for writing CA/keypair's to disk.
+func NewFileBackend(ca *util.CertificateFile, cert *util.CertificateFile, key *util.File) (*FileBackend, error) {
+	if (cert == nil) != (key == nil) {
+		return nil, errors.New("if either cert or key are defined, both fields must be defined as must request to successfully write the keypair to disk")
```
Comment (willbuckner) [inline cert/storage/disk.go]: I can't understand what this error message means, and I imagine users wouldn't understand either :) Not sure what to suggst since I don't quite understand... but I think "both cert and key must be specified if either are specified".

File: certmgr/mgr/file.go [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+type ParsableSpecOptions struct {
+	cert.SpecOptions
+
+	// This defines the service manager to use.  This should be defined
+	// globally rather than per cert- it's allowed here to allow cert
+	// definitions to use a servicemanager of 'command' to allow freeform
+	// invocations.
+	ServiceManagerName string `json:"svcmgr" yaml:"svcmgr"`
+
+	// ServiceManagerTakeActionOnlyIfRunning if set to truee, disables reload/restart attempts
```
Comment (willbuckner) [inline certmgr/mgr/file.go]: nit: `truee`

File: certmgr/mgr/file.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+	if spec.Before == 0 {
+		spec.Before = cert.DefaultBefore
+	}
+	if spec.Interval == 0 {
+		spec.Interval = cert.DefaultInterval
+	}
+
+	err := spec.loadFromPath(path)
+	if err != nil {
+		return nil, err
```
Comment (willbuckner) [inline certmgr/mgr/file.go]: wrapping the error message with some context may be helpful here (and in the rest of this function) as well (but not useful inside `loadFromPath()` itself

File: cert/authority.go (hunk 2 above)
Comment (ferringb) [reply cert/authority.go]: If memory serves, ReadFile already details the path- but I'll wrap the context to clarify that it was while reading this particular CA.

I'll get that shortly.

File: cert/metrics.go (hunk 3 above)
Comment (ferringb) [reply cert/metrics.go]: This metric is actually redundant; we had two metrics, 'write to disk' and 'action notification failed'.

With the new architecture, we still have the 'store failed'- 'store' just now includes "write to disk and trigger an action" if that is the backend in use.

Said another way: for alerting, we only have to track one metric now- not two.  We still have visibility of failure.

File: cert/storage/disk.go (hunk 4 above)
Comment (ferringb) [reply cert/storage/disk.go]: The english on that one is sub-optimal.  It's supposed to error out if you specify just cert or just key- it requires both.

I'll rephrase it shortly.

File: certmgr/mgr/file.go (hunk 5 above)
Comment (ferringb) [reply certmgr/mgr/file.go]: englishh--

File: certmgr/mgr/file.go (hunk 6 above)
Comment (ferringb) [reply certmgr/mgr/file.go]: ack'd, and updating.

File: cert/storage/disk.go (hunk 4 above)
Comment (ferringb) [reply cert/storage/disk.go]: english is updated.  Resolve if there isn't room for improvement please.

File: cert/metrics.go (hunk 3 above)
Comment (willbuckner) [reply cert/metrics.go]: Got it, thanks for the explanation.

File: cert/storage/disk.go (hunk 4 above)
Comment (willbuckner) [reply cert/storage/disk.go]: SGTM :)
